src/go/wsl-helper/pkg/dockerproxy/platform/serve_windows.go

```diff
@@ -0,0 +1,136 @@
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/linuxkit/virtsock/pkg/hvsock"
+)
+
+// DefaultEndpoint is the platform-specific location that dockerd listens on by
+// default.
+const DefaultEndpoint = "npipe:////./pipe/docker_engine"
+
+// ErrListenerClosed is the error that is returned when we attempt to call
+// Accept() on a closed listener.
+var ErrListenerClosed = winio.ErrPipeListenerClosed
+
+// MakeDialer computes the dial function.
+func MakeDialer(port uint32) (func() (net.Conn, error), error) {
+	vmGuid, err := probeVMGUID(port)
+	if err != nil {
+		return nil, fmt.Errorf("could not detect WSL2 VM: %w", err)
+	}
+	dial := func() (net.Conn, error) {
+		conn, err := dialHvsock(vmGuid, port)
+		if err != nil {
+			return nil, err
+		}
+		return conn, nil
+	}
+	return dial, nil
+}
+
+// dialHvsock creates a net.Conn to a Hyper-V VM running Linux with the given
+// GUID, listening on the given vsock port.
+func dialHvsock(vmGuid hvsock.GUID, port uint32) (net.Conn, error) {
+	// go-winio doesn't implement DialHvsock(), but luckily LinuxKit has an
+	// implementation.  We still need go-winio to convert port to GUID.
+	svcGuid, err := hvsock.GUIDFromString(winio.VsockServiceID(port).String())
+	if err != nil {
+		return nil, fmt.Errorf("could not parse Hyper-V service GUID: %w", err)
+	}
+	addr := hvsock.Addr{
+		VMID:      vmGuid,
+		ServiceID: svcGuid,
+	}
+
+	conn, err := hvsock.Dial(addr)
+	if err != nil {
+		return nil, fmt.Errorf("could not dial Hyper-V socket: %w", err)
+	}
+
+	return conn, nil
+}
+
+// Listen on the given Windows named pipe endpoint.
+func Listen(endpoint string) (net.Listener, error) {
+	const prefix = "npipe://"
+
+	if !strings.HasPrefix(endpoint, prefix) {
+		return nil, fmt.Errorf("endpoint %s does not start with protocol %s", endpoint, prefix)
+	}
+
+	listener, err := winio.ListenPipe(endpoint[len(prefix):], nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not listen on %s: %w", endpoint, err)
+	}
+
+	return listener, nil
+}
+
+// ParseBindString parses a HostConfig.Binds entry, returning the (<host-src> or
+// <volume-name>), <container-dest>, and (optional) <options>.  Additionally, it
+// also returns a boolean indicating if the first argument is a host path.
+func ParseBindString(input string) (string, string, string, bool) {
+	// Windows names can be one of a few things:
+	// C:\foo\bar                   colon is possible after the drive letter
+	// \\?\C:\foo\bar               colon is possible after the drive letter
+	// \\server\share\foo           no colons are allowed
+	// \\.\pipe\foo                 no colons are allowed
+	// Luckily, we only have Linux dockerd, so that's easier...
+
+	// pathPattern is a RE for the first two options above.
+	pathPattern := regexp.MustCompile(`^(?:\\\\\?\\)?.:[^:]*`)
+	match := pathPattern.FindString(input)
+	if match == "" {
+		// The first part is a volume name, a pipe, or other non-path thing.
```

Comment:
It doesn't matter — if it's not a path, then we can't munge it (because we don't know how to deal with it); so the only thing we can do is pass to `dockerd` and hopefully it doesn't fail.